### PR TITLE
add opt To allow survey subscription without sending the data on subscribe. 

### PR DESCRIPTION
**Pull request by FredericGuilbault**
**Status**:  Testing 

**Commit list/description**:

This is allow to optin on survey before network is ready. Or just avoid sending a report right at the subscription time.

### CODE REVIEW

**Comment by MichaIng** [comment]
@FredericGuilbault
Many thanks for your PR.

I think how it currently is, this will not work, since INPUT=0 is forced if no settings file was found. But easy to fix.

What we could/should do is remove any call outside of dietpi-update and dietpi-software. This would be doubled anyway since both are called on first login.

Generally DietPi expects network on first boot to do update and software installs. Since it is shipped without any pre-installed software (besides the preinstalled images) it is anyway mostly useless without network then.
And while writing I notice why currently first run update + software is skipped completely on preinstalled images, since those can be "used" OOTB thus do not require network.

Perhaps on first login we should simply do some connectivity test (check network devices for default gateway at least), before calling update and dietpi-software and simply print a recommendation to connect and run update. This would cover as well some other request to allow first boot without network without running into some error messages.

**Comment by FredericGuilbault** [comment]
> Perhaps on first login we should simply do some connectivity test (check network devices for default gateway at least), before calling update and dietpi-software and simply print a recommendation to connect and run update. This would cover as well some other request to allow first boot without network without running into some error messages.

Totally agree. And this would help lysmarine as well. cuz it ship with pre-installed software and updates are not mandatory as I want it functional out of the box without internet. 

It's outside the scope of this PR but I put on my todo list.  ( Im really glad you are so prompt to improvement and moving things around. it's motivating )

**Comment by FredericGuilbault** [comment]
>What we could/should do is remove any call outside of dietpi-update and dietpi-software. This would be doubled anyway since both are called on first login. 

I think data should be send when dietpi-update and dietpi-software are run. But users should be prompted for opt-in on the first boot, we can't really subscribe them (or not) without making them click on some agreement mecanism.

**Comment by MichaIng** [comment]
Okay so basically the agreement is anyway without any effect before the first report is about to be send. It is thus totally okay to ask user not necessarily on first login, but instead when the first report would be send.

The first report makes sense when the first update or install is done and this anyway requires network. And the other way around as fast as network is there, an update makes sense. So:
- Remove network requirement for first boot. Check connection state (via default gateway) on first login. If no connection was found, skip first run update and software and instead print info on login that this is outstanding.
- Skip survey agreement on first login outside of update/installs.
- As fast as network is there on any next login, start firstrun update + installs, which includes the survey agreement prompt and sending.
- Additionally we could add some offline option to dietpi.txt. This would then skip "wait for network" on boot and any network dependant steps on first login automatically, e.g. network time sync and update check + MOTD.

**Comment by FredericGuilbault** [comment]
>Okay so basically the agreement is anyway without any effect before the first report is about to be send.

That is true, But we can't suppose the user will here to answer the question when ever the update process will append. All kinds of things could append. Update could be automated, run in background or on a headless machine, god knows.... This will be specialy true if network is not required during the boot process anymore. 

As it's a blocking mecanism(things can't continue if you don't select)  and it's not requestied by the user ( users will request update and this will prompt) It seem like a receipt for problems. I must say im having a bad feeling about this approach. 
I think the best place for survey opt-in dialog would be right after the GPL agreement dialog

**Comment by MichaIng** [comment]
@FredericGuilbault 
If updates are called without attached terminal (e.g. cron, systemd) or on first login with automated first run setup chosen, G_WHIP prompts are skipped automatically, so no chance it hangs.

The only case where `dietpi-update` is called+applied in background is if auto updates are enabled in dietpi.txt and then it runs as cron job, thus all G_WHIP prompts are skipped. The other way round, if `dietpi-update` is called from within an interactive shell (no chance this happens without user calling it actively or on first login), it requires user actions anyway. Many patch steps prompt G_WHIP info or yes/no questions about important changes or whether to leave packages installed that are not required by DietPi anymore and stuff. So an additional prompt in the end about the survey does not increase any of the concerns you have.

Basically I agree that survey agreement fits best after GPL agreement, but as said that doubles the call on first login and makes things more complex. And again interactive input is mostly required during the update process anyway then.

**Comment by FredericGuilbault** [comment]
Im sorry to came back agan with this, I guess it's the pocces of finding the best solution. 

If G_WHIP is not called before internet is reached. how do we know if the user want to optin or not ? 

---
From what I understand.  In all case, both function `Optin_dialog()`  and `Send_File()` will have to be called on the first boot (or when the internet is reachable ). We can't go out of this. 

The question is, 

* is it better to call it both together when it's time to send the survey. witch imply having `Optin_dialog()` running whenever both G_WHIP and network are available at the same time. 

*  or it;s better to call `Optin_dialog()`on install process and `Send_File()` it's time to send the survey. Witch imply invoking `dietpi-survey` twice.


---


I was thinking of something like this : 
```
	# Usage:

	# - /DietPi/dietpi/dietpi-survey	Ask user to opt in or opt out ( if SURVEY_OPTIN is found in  /DietPi/dietpi.txt this value will be used and the user will not be prompt. )

	# - /DietPi/dietpi/dietpi-survey 1	Non-interactively send survey data or empty file, based on previous user choice
```

So dietPi ask once on first boot with `DietPi/dietpi/dietpi-survey`,   Then every time it need to send the surveydietPi call `DietPi/dietpi/dietpi-survey 1.

**Comment by MichaIng** [comment]
@FredericGuilbault 
If no choice was done yet (settings file does not exist), the survey dialog is forced on every call, regardless of input.
Still, if no terminal is attached or automated firstrun setup is chosen, the dialog is skipped and by default an empty file is sent (opt out state).

Btw I made some changes/simplifications, especially to opt out by default (data should be never sent if user never actively chose to do so). As well the upload file content filled the whole box, so it's moved now into a separate menu option: https://github.com/MichaIng/DietPi/commit/ee695599be36f4080c18da42946dd3731e17aaaf

So lets go through all calls:
- On pre-installed images it is called once during boot: https://github.com/MichaIng/DietPi/blob/dev/dietpi/boot#L122
  - Since this is from a systemd service, it is non-interactive, thus no menu shown, empty file sent by default.
  - However I want to rework pre-installed image handling anyway to run all the default firstrun setup steps. So the above call will be removed, instead first interactive dietpi-update or dietpi-software call will prompt the survey dialog.
- On regular images, any survey dialog is already removed from login script (current dev code, I think we changed this some versions ago), so it is already only called at the end of dietpi-update/software, where network is assured.

So based on the current calls, survey dialog is only called where network is assured to be up. If no choice was done yet and it is called non-interactively, opt out is assumed, on first interactive call, user is asked. So this should be the simplest failsafe solution, or what you think?

**Comment by MichaIng** [comment]
> if SURVEY_OPTIN is found in  /DietPi/dietpi.txt this value will be used and the user will not be prompt.

Ah good idea.

#### EDIT: DietPi-Survey now uses dietpi.txt to read/write opt in/out state: https://github.com/MichaIng/DietPi/commit/1e1722b7f1fb9dfcbcb4e5f4dc9b9f5c80f2315c
- Slightly less performance to scan a larger file via grep/sed, but compared to the upload time is is not measurable.
- Benefit is that user can opt in/out prior to first boot to further automate firstrun setup.
- Also 1 file less means 4k less RAMdisk size to handle

**Comment by MichaIng** [comment]
@FredericGuilbault 
Okay I close this PR, as I think current implementation resolves your concerns.

The other topic about offline mode is something I will do with v6.26 I think since I want to release v6.25 this month (including beta phase) and this requires a bid more testing: https://github.com/MichaIng/DietPi/issues/2915

**Comment by FredericGuilbault** [comment]
Good. I will take a look in that.

**Comment by FredericGuilbault** [comment]
I have made a review of your commit, it seem good to me. 

On an other topic it seem strange to me to send empty files when it's opted out.  
No means no XD.

But that's not a bug it's a design choice.

**Comment by MichaIng** [comment]
@FredericGauilbault
Yeah, the first new survey implementation some versions back had three options:
- Opt in
- Opt out
- Opt out and purge any uploaded data

The problem is that the upload user that is used does not have any read or file list access to the upload directory. The generated unique ID that defines the uploaded file is at least, since unique, something that can be associated with a certain machine, although one needs direct access to that machine to find it which allows to gather the same data as well. However we decided that no one should be able to have any read access to the upload dir to be fully failsafe, also if in case of an error some unintended data is written to the file. And because of that the upload user can as well not remove any files, it can only overwrite it with an empty file to purge content.
This also makes sense to prevent attackers from emptying our survey dir. Without file list access it is nearly impossible to by chance guess/hit one of the unique IDs for purging. And invalid accesses are banned by fail2ban.

And finally to make the menu and handling simpler, now any uploaded data is simply always purged when opting out by uploading an empty file.

One benefit for us is that we have an overall count of DietPi systems and can see how many users decide to opt in vs out.

But yeah would be best actually best to somehow only upload an empty file once, when also data was uploaded before, but this would require e.g. a flag file again. And yeah the benefit of no upload vs empty upload is not really there, during installs and updates the server is anyway accessed with much heavier access.
